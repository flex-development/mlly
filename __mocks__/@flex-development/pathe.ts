/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
