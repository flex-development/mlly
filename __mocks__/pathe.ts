/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

/**
 * [`pathe`][1] module type.
 *
 * [1]: https://github.com/unjs/pathe
 */
type Actual = typeof import('pathe')

/**
 * `pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const isAbsolute = vi.fn(actual.isAbsolute)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)
