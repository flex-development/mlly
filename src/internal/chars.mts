/**
 * @file Internal - chars
 * @module mlly/internal/chars
 */

/**
 * Characters that may have special meaning.
 *
 * @enum {string}
 */
enum chars {
  asterisk = '*',
  at = '@',
  empty = '',
  hash = '#'
}

export default chars
