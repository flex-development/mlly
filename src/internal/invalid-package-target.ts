/**
 * @file Internal - invalidPackageTarget
 * @module mlly/internal/invalidPackageTarget
 */

import chars from '#internal/chars'
import {
  ERR_INVALID_PACKAGE_TARGET,
  type ErrInvalidPackageTarget
} from '@flex-development/errnode'
import type { ModuleId } from '@flex-development/mlly'
import { fileURLToPath } from '@flex-development/pathe'

/**
 * Create an {@linkcode ERR_INVALID_PACKAGE_TARGET} error.
 *
 * @see {@linkcode ErrInvalidPackageTarget}
 * @see {@linkcode ModuleId}
 *
 * @internal
 *
 * @param {ModuleId} packageUrl
 *  URL of directory containing `package.json` file
 * @param {string} subpath
 *  Subpath pattern
 * @param {unknown} target
 *  The invalid package target
 * @param {boolean | null | undefined} [isImports]
 *  Whether `target` is internal to the package
 * @param {ModuleId | null | undefined} [parent]
 *  URL of parent module
 * @return {ErrInvalidPackageTarget}
 *  New node error
 */
function invalidPackageTarget(
  packageUrl: ModuleId,
  subpath: string,
  target: unknown,
  isImports?: boolean | null | undefined,
  parent?: ModuleId | null | undefined
): ErrInvalidPackageTarget {
  return new ERR_INVALID_PACKAGE_TARGET(
    fileURLToPath(packageUrl),
    subpath,
    typeof target === 'object' && target !== null
      ? JSON.stringify(target, null, '')
      : String(target) + (typeof target === 'bigint' ? 'n' : chars.empty),
    isImports,
    parent && fileURLToPath(parent)
  )
}

export default invalidPackageTarget
