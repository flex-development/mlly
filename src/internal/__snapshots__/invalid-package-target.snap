// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:internal/invalidPackageTarget > should return new \`ERR_INVALID_PACKAGE_TARGET\` (0) 1`] = `Invalid "imports" target "./__fixtures__/node_modules/*" defined for '#nm/*' in the package config \${process.cwd()}/package.json imported from \${process.cwd()}/src/internal/__tests__/invalid-package-target.spec.mts`;

exports[`unit:internal/invalidPackageTarget > should return new \`ERR_INVALID_PACKAGE_TARGET\` (1) 1`] = `Invalid "exports" main target "[null]" defined in the package config \${process.cwd()}/__fixtures__/node_modules/invalid-target-array/package.json; targets must start with "./"`;

exports[`unit:internal/invalidPackageTarget > should return new \`ERR_INVALID_PACKAGE_TARGET\` (2) 1`] = `Invalid "exports" main target "13n" defined in the package config \${process.cwd()}/__fixtures__/node_modules/invalid-target-bigint/package.json; targets must start with "./"`;
